keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -1539,63 +1610,85 @@ type ResourceClaimParameters struct {
     // by multiple consumers at the same time.
     Shareable bool
 
-    // DriverRequests describes all resources that are needed for the
-    // allocated claim. A single claim may use resources coming from
-    // different drivers. For each driver, this array has at most one
-    // entry which then may have one or more per-driver requests.
-    //
-    // May be empty, in which case the claim can always be allocated.
-    DriverRequests []DriverRequests
-}
-
-// DriverRequests describes all resources that are needed from one particular driver.
-type DriverRequests struct {
-    // DriverName is the name used by the DRA driver kubelet plugin.
-    DriverName string
-
-    // VendorParameters are arbitrary setup parameters for all requests of the
-    // claim. They are ignored while allocating the claim.
-    VendorParameters runtime.Object
-
-    // Requests describes all resources that are needed from the driver.
-    Requests []ResourceRequest
-}
-
-// ResourceRequest is a request for resources from one particular driver.
-type ResourceRequest struct {
-    // VendorParameters are arbitrary setup parameters for the requested
-    // resource. They are ignored while allocating a claim.
-    VendorParameters runtime.Object
+    // ClaimConfig defines configuration parameters that apply to the entire claim.
+    // They are ignored while allocating the claim.
+    Config *ConfigurationParameters
 
     ResourceRequestModel
 }
 
 // ResourceRequestModel must have one and only one field set.
 type ResourceRequestModel struct {
     // NamedResources describes a request for resources with the named resources model.
-    NamedResources *NamedResourcesRequest
+    // The list may be empty, in which case the claim can be allocated without
+    // using any instances.
+    NamedResources *[]NamedResourcesRequest
```

Comment:
Argh!

The gogo protobuf code generator fails for this. It generates `len(NamedResources)` which does not compile because `NamedResources` is a pointer. Not sure yet how to resolve this.

I can wrap `[]NamedResourcesRequest` in a struct, but the the YAML and JSON end up with one additional indirection.

This is not a show-stopper for "vendor-independent attributes", but it throws a monkey wrench into the attempt to flatten the YAML at the same time.

Reply:
How does this look?

```yaml

requests:
- config:
    vendor:
      driverName: cards.dra.example.com
      parameters: <per-request parameters>
  device:
     selector: <CEL expression for named resources>
```

Reply:
Or

```yaml
request:
- device:
    selector: <CEL expression for named resources>
  config:
    vendor:
      driverName: cards.dra.example.com
      parameters: <per-request parameters>
```

Reply:
Both is assuming that "named resources" gets renamed to "device". For now, it would have "namedResource".

Reply:
Do we have a convention to use plural form for lists?

I'm wondering about `request` (imperative) vs `requests` (noun, plural).